File: rust/src/parse_pattern.rs
use crate::token::{Token, TokenOptions};
use std::collections::HashMap;

pub type Dictionaries = HashMap<String, Vec<String>>;

fn is_special(c: char) -> bool {
    matches!(c, '#' | '@' | '$' | '*' | '&' | '?' | '!' | '-' | '%')
}

/// Split like JS/Python capturing-group split, without a regex crate.
fn split_keeping_delimiters(input: &str) -> Vec<String> {
    let chars: Vec<char> = input.chars().collect();
    let mut parts = Vec::new();
    let mut i = 0;
    let mut literal_start = 0;

    while i < chars.len() {
        // Escaped special: \X
        if chars[i] == '\\' && i + 1 < chars.len() && is_special(chars[i + 1]) {
            if i > literal_start {
                parts.push(chars[literal_start..i].iter().collect());
            }
            parts.push(chars[i..=i + 1].iter().collect());
            i += 2;
            literal_start = i;
            continue;
        }

        // Special with {…}
        if is_special(chars[i]) && i + 1 < chars.len() && chars[i + 1] == '{' {
            if i > literal_start {
                parts.push(chars[literal_start..i].iter().collect());
            }
            let mut j = i + 2;
            while j < chars.len() && chars[j] != '}' {
                j += 1;
            }
            if j < chars.len() && chars[j] == '}' {
                parts.push(chars[i..=j].iter().collect());
                i = j + 1;
                literal_start = i;
                continue;
            }
        }

        // Bare special
        if is_special(chars[i]) {
            if i > literal_start {
                parts.push(chars[literal_start..i].iter().collect());
            }
            parts.push(chars[i].to_string());
            i += 1;
            literal_start = i;
            continue;
        }

        i += 1;
    }

    if literal_start < chars.len() {
        parts.push(chars[literal_start..].iter().collect());
    }
    parts
}

fn parse_length_with_variants(part: &str, variants: Vec<String>) -> TokenOptions {
    let mut start_length = 1;
    let mut end_length = 1;

    if let Some(open) = part.find('{') {
        if let Some(close) = part[open..].find('}') {
            let inner = &part[open + 1..open + close];
            if let Some(dash) = inner.find('-') {
                if let (Ok(s), Ok(e)) = (
                    inner[..dash].parse::<i32>(),
                    inner[dash + 1..].parse::<i32>(),
                ) {
                    start_length = s;
                    end_length = e;
                }
            } else if let Ok(n) = inner.parse::<i32>() {
                start_length = n;
                end_length = n;
            }
        }
    }

    TokenOptions {
        string: None,
        start_length: Some(start_length),
        end_length: Some(end_length),
        variants,
        src: part.to_string(),
    }
}

fn parse_length_with_string(part: &str) -> Option<TokenOptions> {
    // Match {'...'} with optional ,N-M or ,N — greedy content between quotes.
    let open = part.find("{'")?;
    let after_open = open + 2;
    let rest = &part[after_open..];
    let close_quote = rest.rfind('\'')?;
    let content = &rest[..close_quote];
    let after_quote = &rest[close_quote + 1..];

    if !after_quote.starts_with('}') && !after_quote.starts_with(',') {
        // Must end with } eventually
        if !after_quote.contains('}') {
            return None;
        }
    }

    let mut start_length = 1;
    let mut end_length = 1;

    if let Some(stripped) = after_quote.strip_prefix(',') {
        let before_brace = stripped.strip_suffix('}').unwrap_or(stripped);
        if let Some(dash) = before_brace.find('-') {
            if let (Ok(s), Ok(e)) = (
                before_brace[..dash].parse::<i32>(),
                before_brace[dash + 1..].parse::<i32>(),
            ) {
                start_length = s;
                end_length = e;
            }
        } else if let Ok(n) = before_brace.parse::<i32>() {
            start_length = n;
            end_length = n;
        }
    } else if !after_quote.starts_with('}') {
        return None;
    }

    Some(TokenOptions {
        string: Some(content.to_string()),
        start_length: Some(start_length),
        end_length: Some(end_length),
        variants: Vec::new(),
        src: part.to_string(),
    })
}

fn chars_as_variants(s: &str) -> Vec<String> {
    s.chars().map(|c| c.to_string()).collect()
}

fn simple_tokenizer(part: &str, alphabet: &str) -> Token {
    Token::new(parse_length_with_variants(part, chars_as_variants(alphabet)))
}

fn dictionary_tokenizer(part: &str, dictionaries: &Dictionaries) -> Token {
    match parse_length_with_string(part) {
        Some(mut options)
            if options
                .string
                .as_ref()
                .map(|s| s.is_empty() || dictionaries.contains_key(s))
                .unwrap_or(false) =>
        {
            let key = options.string.clone().unwrap_or_default();
            options.variants = dictionaries.get(&key).cloned().unwrap_or_default();
            Token::new(options)
        }
        _ => Token::new(TokenOptions {
            string: None,
            start_length: Some(1),
            end_length: Some(1),
            variants: vec![part.to_string()],
            src: part.to_string(),
        }),
    }
}

fn words_tokenizer(part: &str) -> Token {
    match parse_length_with_string(part) {
        None => Token::new(TokenOptions {
            string: None,
            start_length: Some(1),
            end_length: Some(1),
            variants: vec![part.to_string()],
            src: part.to_string(),
        }),
        Some(mut options) => {
            let mut variants = Vec::new();
            let mut work_string = options.string.clone().unwrap_or_default();
            let mut index = 0;
            while index < work_string.len() {
                if index + 1 < work_string.len()
                    && work_string.as_bytes()[index] == b'\\'
                    && work_string.as_bytes()[index + 1] == b','
                {
                    index += 2;
                } else if work_string.as_bytes()[index] == b',' {
                    variants.push(work_string[..index].to_string());
                    work_string = work_string[index + 1..].to_string();
                    index = 0;
                } else {
                    index += 1;
                }
            }
            variants.push(work_string);
            options.variants = variants
                .into_iter()
                .map(|v| v.replace("\\,", ","))
                .collect();
            Token::new(options)
        }
    }
}

fn part_to_token(part: &str, dictionaries: &Dictionaries) -> Token {
    let first = part.chars().next();
    match first {
        Some('#') => simple_tokenizer(part, "0123456789"),
        Some('@') => simple_tokenizer(part, "abcdefghijklmnopqrstuvwxyz"),
        Some('*') => simple_tokenizer(part, "abcdefghijklmnopqrstuvwxyz0123456789"),
        Some('-') => simple_tokenizer(
            part,
            "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789",
        ),
        Some('!') => simple_tokenizer(part, "ABCDEFGHIJKLMNOPQRSTUVWXYZ"),
        Some('?') => simple_tokenizer(part, "ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"),
        Some('&') => {
            simple_tokenizer(part, "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ")
        }
        Some('%') => dictionary_tokenizer(part, dictionaries),
        Some('$') => words_tokenizer(part),
        Some('\\') if part.len() > 1 && is_special(part.chars().nth(1).unwrap()) => {
            Token::new(TokenOptions {
                string: None,
                start_length: Some(1),
                end_length: Some(1),
                variants: vec![part[1..].to_string()],
                src: part.to_string(),
            })
        }
        _ => Token::new(TokenOptions {
            string: None,
            start_length: Some(1),
            end_length: Some(1),
            variants: vec![part.to_string()],
            src: part.to_string(),
        }),
    }
}

pub fn parse_pattern(input_pattern: &str, dictionaries: &Dictionaries) -> Vec<Token> {
    split_keeping_delimiters(input_pattern)
        .into_iter()
        .filter(|p| !p.is_empty())
        .map(|p| part_to_token(&p, dictionaries))
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn splits_year_digits() {
        let parts = split_keeping_delimiters("Year 19##");
        assert_eq!(parts, vec!["Year 19", "#", "#"]);
    }
}

File: r/R/parse_pattern.R
SPECIAL_CHARS <- c("#", "@", "$", "*", "&", "?", "!", "-", "%")

is_special <- function(c) {
  c %in% SPECIAL_CHARS
}

chars_as_variants <- function(s) {
  if (nchar(s) == 0L) {
    return(character(0))
  }
  strsplit(s, "", fixed = TRUE)[[1]]
}

split_keeping_delimiters <- function(input) {
  if (nchar(input) == 0L) {
    return(character(0))
  }

  parts <- character(0)
  i <- 1L
  literal_start <- 1L
  len <- nchar(input)

  while (i <= len) {
    c <- substr(input, i, i)

    if (c == "\\" && i + 1L <= len && is_special(substr(input, i + 1L, i + 1L))) {
      if (i > literal_start) {
        parts <- c(parts, substr(input, literal_start, i - 1L))
      }
      parts <- c(parts, substr(input, i, i + 1L))
      i <- i + 2L
      literal_start <- i
    } else if (is_special(c) && i + 1L <= len && substr(input, i + 1L, i + 1L) == "{") {
      if (i > literal_start) {
        parts <- c(parts, substr(input, literal_start, i - 1L))
      }
      j <- i + 2L
      while (j <= len && substr(input, j, j) != "}") {
        j <- j + 1L
      }
      if (j <= len && substr(input, j, j) == "}") {
        parts <- c(parts, substr(input, i, j))
        i <- j + 1L
        literal_start <- i
      } else {
        if (i > literal_start) {
          parts <- c(parts, substr(input, literal_start, i - 1L))
        }
        parts <- c(parts, c)
        i <- i + 1L
        literal_start <- i
      }
    } else if (is_special(c)) {
      if (i > literal_start) {
        parts <- c(parts, substr(input, literal_start, i - 1L))
      }
      parts <- c(parts, c)
      i <- i + 1L
      literal_start <- i
    } else {
      i <- i + 1L
    }
  }

  if (literal_start <= len) {
    parts <- c(parts, substr(input, literal_start, len))
  }

  parts
}

parse_length_with_variants <- function(part, variants) {
  start_length <- 1L
  end_length <- 1L

  open <- regexpr("{", part, fixed = TRUE)[1]
  if (open > 0L) {
    close <- regexpr("}", part, fixed = TRUE)[1]
    if (close > open) {
      inner <- substr(part, open + 1L, close - 1L)
      dash <- regexpr("-", inner, fixed = TRUE)[1]
      if (dash > 0L) {
        s <- suppressWarnings(as.integer(substr(inner, 1L, dash - 1L)))
        e <- suppressWarnings(as.integer(substr(inner, dash + 1L, nchar(inner))))
        if (!is.na(s) && !is.na(e)) {
          start_length <- s
          end_length <- e
        }
      } else {
        n <- suppressWarnings(as.integer(inner))
        if (!is.na(n)) {
          start_length <- n
          end_length <- n
        }
      }
    }
  }

  list(
    variants = variants,
    startLength = start_length,
    endLength = end_length,
    src = part
  )
}

parse_length_with_string <- function(part) {
  open <- regexpr("{'", part, fixed = TRUE)[1]
  if (open <= 0L) {
    return(FALSE)
  }

  after_open <- open + 2L
  rest <- substr(part, after_open, nchar(part))
  close_quote <- 0L
  for (idx in seq(nchar(rest), 1L, by = -1L)) {
    if (substr(rest, idx, idx) == "'") {
      close_quote <- idx
      break
    }
  }
  if (close_quote == 0L) {
    return(FALSE)
  }

  content <- substr(rest, 1L, close_quote - 1L)
  after_quote <- substr(rest, close_quote + 1L, nchar(rest))

  if (!startsWith(after_quote, "}") && !startsWith(after_quote, ",")) {
    if (regexpr("}", after_quote, fixed = TRUE)[1] <= 0L) {
      return(FALSE)
    }
  }

  start_length <- 1L
  end_length <- 1L

  if (startsWith(after_quote, ",")) {
    stripped <- substr(after_quote, 2L, nchar(after_quote))
    if (endsWith(stripped, "}")) {
      stripped <- substr(stripped, 1L, nchar(stripped) - 1L)
    }
    dash <- regexpr("-", stripped, fixed = TRUE)[1]
    if (dash > 0L) {
      s <- suppressWarnings(as.integer(substr(stripped, 1L, dash - 1L)))
      e <- suppressWarnings(as.integer(substr(stripped, dash + 1L, nchar(stripped))))
      if (!is.na(s) && !is.na(e)) {
        start_length <- s
        end_length <- e
      }
    } else {
      n <- suppressWarnings(as.integer(stripped))
      if (!is.na(n)) {
        start_length <- n
        end_length <- n
      }
    }
  } else if (!startsWith(after_quote, "}")) {
    return(FALSE)
  }

  list(
    string = content,
    startLength = start_length,
    endLength = end_length,
    src = part
  )
}

simple_tokenizer <- function(variants_string) {
  variants <- chars_as_variants(variants_string)
  function(part) {
    create_token(parse_length_with_variants(part, variants))
  }
}

dictionary_tokenizer <- function(part, dictionaries) {
  options <- parse_length_with_string(part)
  if (isFALSE(options) ||
      (!is.null(options$string) && options$string != "" &&
       is.null(dictionaries[[options$string]]))) {
    options <- list(
      variants = part,
      startLength = 1L,
      endLength = 1L,
      src = part
    )
  } else {
    key <- if (is.null(options$string)) "" else options$string
    options$variants <- if (is.null(dictionaries[[key]])) character(0) else dictionaries[[key]]
  }
  create_token(options)
}

words_tokenizer <- function(part) {
  options <- parse_length_with_string(part)

  if (isFALSE(options)) {
    options <- list(
      variants = part,
      startLength = 1L,
      endLength = 1L,
      src = part
    )
  } else {
    variants <- character(0)
    work_string <- if (is.null(options$string)) "" else options$string
    index <- 1L
    while (index <= nchar(work_string)) {
      if (substr(work_string, index, index + 1L) == "\\,") {
        index <- index + 2L
      } else if (substr(work_string, index, index) == ",") {
        variants <- c(variants, substr(work_string, 1L, index - 1L))
        work_string <- substr(work_string, index + 1L, nchar(work_string))
        index <- 1L
      } else {
        index <- index + 1L
      }
    }
    variants <- c(variants, work_string)
    options$variants <- gsub("\\,", ",", variants, fixed = TRUE)
  }

  create_token(options)
}

part_to_token <- function(part, dictionaries) {
  tokenizers <- list(
    "#" = simple_tokenizer("0123456789"),
    "@" = simple_tokenizer("abcdefghijklmnopqrstuvwxyz"),
    "*" = simple_tokenizer("abcdefghijklmnopqrstuvwxyz0123456789"),
    "-" = simple_tokenizer("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"),
    "!" = simple_tokenizer("ABCDEFGHIJKLMNOPQRSTUVWXYZ"),
    "?" = simple_tokenizer("ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"),
    "&" = simple_tokenizer("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"),
    "%" = function(p) dictionary_tokenizer(p, dictionaries),
    "$" = words_tokenizer
  )

  first <- if (nchar(part) > 0L) substr(part, 1L, 1L) else ""
  tokenizer <- tokenizers[[first]]
  is_escaped <- nchar(part) > 1L &&
    substr(part, 1L, 1L) == "\\" &&
    !is.null(tokenizers[[substr(part, 2L, 2L)]])

  if (!is.null(tokenizer)) {
    return(tokenizer(part))
  }
  if (is_escaped) {
    return(create_token(list(
      variants = substr(part, 2L, nchar(part)),
      src = part
    )))
  }
  create_token(list(
    variants = part,
    src = part
  ))
}

parse_pattern <- function(input_pattern, dictionaries = list()) {
  if (is.null(dictionaries)) {
    dictionaries <- list()
  }
  parts <- split_keeping_delimiters(input_pattern)
  tokens <- lapply(parts[parts != ""], function(part) part_to_token(part, dictionaries))
  tokens
}
